### PR TITLE
Add batch files to execute Dotty commands on Windows

### PR DESCRIPTION
**Pull request by michelou**
1) Batch files present in directory [**`dist\bin\`**](https://github.com/lampepfl/dotty/tree/master/dist/bin) will be copied unchanged to the gz/zip archives (see `sbt dist-bootstrapped/packArchive`).
2) ~~Batch files added to [**`project\scripts\`**](https://github.com/lampepfl/dotty/tree/master/project/scripts) (ie. `build.bat`, `cmdTests.bat` and `bootstrapCmdTests.bat`) are adapted from [`.drone.yml`](https://github.com/lampepfl/dotty/blob/master/.drone.yml) (executed on the [Dotty CI](http://dotty-ci.epfl.ch/lampepfl/dotty) server) and can generate the documentation and the gz/zip archives (*depending* on bootstrap and tests results).~~

**NB.** Point 1 is ready to be tested on a Windows machine, ~~but point 2 requires my changes in **`compiler\src\*.scala`** and **`compiler\test\*.scala`** to be committed in a new branch (and submitted as a separate PR).~~ Points 1 and 2 have been tested on a Windows machine and are used in PR #5531.

### CODE REVIEW

**Comment by liufengyun** [comment]
Thanks a lot @michelou !

I created a test here: https://github.com/lampepfl/packtest/pull/18

It seems there is some error when running the batch files. As I've very little experience with batch scripts. Could you please have a look? Thanks a lot.

https://ci.appveyor.com/project/liufengyun/packtest-s6xq2/builds/20551431/job/mdakdmt9m9pp6jlr

**Comment by liufengyun** [comment]
Thanks @michelou , the CI now is green: https://github.com/lampepfl/packtest/pull/18

File: project/scripts/build.bat [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+echo     cleanall               clean project (sbt+git) and quit
+echo     clone                  update submodules
+echo     compile                generate compiler 1st stage (after clone)
+echo     doc[umentation]        generate documentation (after bootstrap)
+echo     help                   display this help message
+echo   Advanced subcommands (no deps):
+echo     arch[ives]-only        generate ONLY gz/zip archives
+echo     boot[strap]-only       generate ONLY compiler bootstrap
+echo     compile-only           generate ONLY compiler 1st stage
+echo     doc[umentation]-only]  generate ONLY documentation
```
Comment (liufengyun) [inline project/scripts/build.bat]: What is the motivation to have both tests and some utility in the same batch?

File: project/scripts/build.bat [hunk 2]
```diff
@@ -0,0 +1,458 @@
+@echo off
+setlocal enabledelayedexpansion
+
```
Comment (liufengyun) [inline project/scripts/build.bat]: My only concern with this file is that we need CI for windows. Related issue: #4588 

We can try to integrate `appveyor` and run `build.bat`, but NOT the whole test set.

**Comment by michelou** [comment]
I wrote `projets/scripts/build.bat` to mimic the phases executed on [DottyIC](http://dotty-ci.epfl.ch/lampepfl/dotty) on a Windows workstation **and not** on an AppVeyor server.

**Comment by michelou** [comment]
After consideration I will move that file to my personal Git account. The more general question is : where do such build scripts (both for Unix and Windows) belong to ?

**Comment by liufengyun** [comment]
@michelou Thanks for the clarification. My only concern is about maintainability -- if we don't run `project/scripts/build.bat` in some CI, then it's highly likely it will break unnoticed in the long run. If we can run some check on windows, it's always a win regardless of the coverage (no need to be the same as `.drone.yml`).

**Comment by smarter** [comment]
@liufengyun Note that apparently the latest version of drone supports Windows, and we'd like to add Windows to our CI eventually.

**Comment by michelou** [comment]
Short answer : `build help`
Long answer: [Building Dotty on Windows](https://github.com/michelou/dotty-examples/blob/master/DRONE.md)
So you choosed to call `project\scripts\build.bat` instead of porting the logic `.drone.yml` to `.appveyor` (both ways have advantages/disadvantages).

File: project/scripts/bootstrapCmdTests.bat [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+rem ## Main
+
+rem # check that benchmarks can run
+if %_DEBUG%==1 echo [%_BASENAME%] "%_SBT_CMD%" "dotty-bench/jmh:run 1 1 tests/pos/alias.scala"
+call "%_SBT_CMD%" "dotty-bench/jmh:run 1 1 tests/pos/alias.scala"
+if not %ERRORLEVEL%==0 ( set _EXITCODE=1& goto end )
+
+rem # The above is here as it relies on the bootstrapped library.
+call "%_SBT_CMD%" "dotty-bench-bootstrapped/jmh:run 1 1 tests/pos/alias.scala"
+call "%_SBT_CMD%" "dotty-bench-bootstrapped/jmh:run 1 1 -with-compiler compiler/src/dotty/tools/dotc/core/Types.scala"
```
Comment (liufengyun) [inline project/scripts/bootstrapCmdTests.bat]: One naive question: we don't need to use windows separator for source file, like `tests\pos\alias.scala`?

File: project/scripts/bootstrapCmdTests.bat [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+echo testing scala.quoted.Expr.run from sbt dotr
+call "%_SBT_CMD%" ";dotty-compiler-bootstrapped/dotc tests/run-with-compiler/quote-run.scala; dotty-compiler-bootstrapped/dotr -with-compiler Test" > "%_TMP_FILE%"
+call :grep "val a: scala.Int = 3" "%_TMP_FILE%"
+if not %_EXITCODE%==0 goto end
```
Comment (liufengyun) [inline project/scripts/bootstrapCmdTests.bat]: It seems the error handling is not effective: https://ci.appveyor.com/project/liufengyun/dotty/builds/20722266#L1413

The script continues to execute and exits normally.

File: project/scripts/bootstrapCmdTests.bat (hunk 3 above)
Comment (michelou) [reply project/scripts/bootstrapCmdTests.bat]: Path `tests/pos/alias.scala` is given **inside** a sbt command wich supports both kinds of file separators.

File: project/scripts/bootstrapCmdTests.bat (hunk 4 above)
Comment (michelou) [reply project/scripts/bootstrapCmdTests.bat]: Correct. I wrote parts of `bootstrapCmdTests.bat` after 1AM in the morning and I am aware that some changes are still necessary.

**Comment by odersky** [comment]
@liufengyun What's the status here?

**Comment by liufengyun** [comment]
@odersky Just synced with @michelou . There are still 3 language server tests that fail on Windows. After that we will need to get a dedicated Windows machine, and upgrade Drone to support windows.

**Comment by michelou** [comment]
Batch files are still available from my project [`michelou/dotty-examples`](https://github.com/michelou/dotty-examples) (see file [`BUILD.md`](https://github.com/michelou/dotty-examples/blob/master/BUILD.md)).

**Pull request closed without merging**